app/src/behavior.c

```diff
@@ -39,6 +41,144 @@ const struct device *z_impl_behavior_get_binding(const char *name) {
+        if (usage_id >
+            COND_CODE_1(IS_ENABLED(CONFIG_ZMK_HID_CONSUMER_REPORT_USAGES_BASIC), (0xFF), (0xFFF))) {
+            return -EINVAL;
+        }
+        break;
+    default:
+        LOG_WRN("Unsupported HID usage page %d", usage_page);
+        return -EINVAL;
+    }
+
+    return 0;
+}
+
+int zmk_behavior_validate_param_values(const struct behavior_parameter_value_metadata *values,
+                                       size_t values_len, uint32_t param) {
+    bool found_match = false;
+
+    if (values_len == 0) {
+        return -ENODEV;
+    }
+
+    for (int v = 0; v < values_len; v++) {
+        const struct behavior_parameter_value_metadata *value_meta = &values[v];
+
+        switch (value_meta->type) {
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_NIL:
+            if (param == 0) {
+                found_match = true;
+            }
+            break;
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_HID_USAGE:
+            if (validate_hid_usage(ZMK_HID_USAGE_PAGE(param), ZMK_HID_USAGE_ID(param)) >= 0) {
+                found_match = true;
+            }
+
+            break;
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_LAYER_INDEX:
+            if (param >= 0 && param < ZMK_KEYMAP_LEN) {
+                found_match = true;
+            }
+            break;
+            /* TODO: Restore with HSV -> RGB refactor
+            case BEHAVIOR_PARAMETER_STANDARD_DOMAIN_HSV:
+                // TODO: No real way to validate? Maybe max brightness?
+                break;
+            */
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_VALUE:
+            if (param == value_meta->value) {
+                found_match = true;
+            }
+            break;
+        case BEHAVIOR_PARAMETER_VALUE_TYPE_RANGE:
+            if (param >= value_meta->range.min && param <= value_meta->range.max) {
+                found_match = true;
+            }
+            break;
+        }
+
+        if (found_match) {
+            break;
+        }
+    }
+
+    return found_match ? 0 : -ENOTSUP;
+}
+
+int zmk_behavior_validate_params_metadata(const struct behavior_parameter_metadata *metadata,
+                                          uint32_t param1, uint32_t param2) {
+    if (!metadata || metadata->sets_len == 0) {
+        if (!metadata) {
+            LOG_ERR("No metadata to check against");
+        }
+
+        return (param1 == 0 && param2 == 0) ? 0 : -ENODEV;
+    }
+
+    for (int s = 0; s < metadata->sets_len; s++) {
+        const struct behavior_parameter_metadata_set *set = &metadata->sets[s];
+        int param1_ret =
+            zmk_behavior_validate_param_values(set->param1_values, set->param1_values_len, param1);
+        int param2_ret =
+            zmk_behavior_validate_param_values(set->param2_values, set->param2_values_len, param2);
+
+        if ((param1_ret >= 0 || (param1_ret == -ENODEV && param1 == 0)) &&
+            (param2_ret >= 0 || (param2_ret == -ENODEV && param2 == 0))) {
+            LOG_DBG("ALL GOOD!");
```

Comment:
Probably should either remove this or add some extra context (what did it correctly match?) so the log message is useful later.